### PR TITLE
:book: multiple version tutorial

### PR DESCRIPTION
**Pull request by droot**
Main highlights:
 - Tutorial describing how to implement multiple versions using Kubebuilder
 - An example project with multiple versions 

You can preview this tutorial [here](https://deploy-preview-851--kubebuilder.netlify.com/multiversion-tutorial/tutorial.html)

### CODE REVIEW

File: docs/book/src/multiversion-tutorial/testdata/project/api/v1/disk_types.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+/*
+We need to specify the version that is being used as storage version. In this
+case, we decided to use v1 version for storage, so we use crd marker `+kubebuilder:storageversion`
+on v1.Disk type to indicate that.
+*/
+
+// +kubebuilder:object:root=true
+
+// +kubebuilder:storageversion <-- not the storageversion marker
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/api/v1/disk_types.go]: > not the storageversion marker

This confuses me.

File: docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	Price Price `json:"price"`
+}
+
+// Price represents a generic price value that has amount and currency.
+type Price struct {
+	Amount   int64  `json:"amount"`
+	Currency string `json:"currency"`
+}
+
+// DiskStatus defines the observed state of Disk
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go]: We can collapse from `DiskStatus` to `type Disk struct` to focus on what really matter.

File: docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+			Amount:   int64(amount),
+			Currency: parts[1],
+		}
+		return nil
+	default:
+		return fmt.Errorf("unsupported type %v", t)
+	}
+}
+
+// +kubebuilder:object:root=true
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go]: Collapse the DiskList and init.

File: docs/book/src/multiversion-tutorial/testdata/project/api/v3/disk_types.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	PricePerGB Price `json:"pricePerGB"`
+}
+
+// Price represents a generic price value that has amount and currency.
+type Price struct {
+	Amount   int64  `json:"amount"`
+	Currency string `json:"currency"`
+}
+
+// DiskStatus defines the observed state of Disk
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/api/v3/disk_types.go]: We can collapse from `DiskStatus` to `type Disk struct`.

File: docs/book/src/multiversion-tutorial/testdata/project/api/v3/disk_types.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+			Amount:   int64(amount),
+			Currency: parts[1],
+		}
+		return nil
+	default:
+		return fmt.Errorf("unsupported type %v", t)
+	}
+}
+
+// +kubebuilder:object:root=true
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/api/v3/disk_types.go]: Collapse DiskList and init

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+have versions v1, v2 and v3 of a kind, kubernetes will use one of the versions to
+persist the object of that Kind in Etcd. You can specify the version to be used
+for storage in the Custom Resource definition for that API.
+
+One can think of storage version as the hub and other versions as spokes to visualize the
+relationship between storage and other versions (as shown below in the diagram).
+The important thing to note is that conversion between storage and other version
+should be lossless (round trippable). As shown in the diagram below, v3 is the 
+storage/hub version and v1 and v2 spoke version. This tutorial uses 
+the terms `storage version` and `hub` interchangeably.
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/tutorial.md]: We should explain briefly how does converting spoke to spoke works.

File: docs/book/src/multiversion-tutorial/testdata/project/config/default/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  # manager_prometheus_metrics_patch.yaml should be enabled.
+#- manager_prometheus_metrics_patch.yaml
+
+# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix including the one in crd/kustomization.yaml
+- manager_webhook_patch.yaml
+
+# [CAINJECTION] Uncomment next line to enable the CA injection in the admission webhooks.
+# Uncomment 'CAINJECTION' in crd/kustomization.yaml to enable the CA injection in the admission webhooks.
+# 'CERTMANAGER' needs to be enabled to use ca injection
+#- webhookcainjection_patch.yaml
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/testdata/project/config/default/kustomization.yaml]: Stuff under `config` needs to be rebased to pick up https://github.com/kubernetes-sigs/kubebuilder/pull/852

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+- Ensure that `patches/webhook_in_<kind>.yaml` and `patches/cainjection_in_<kind>.yaml` are enabled in `config/crds/kustomization.yaml` file 
+- Ensure that `../certmanager` and `../webhook` directories are enabled under `bases` section in `config/default/kustomization.yaml` file
+- Ensure that `manager_webhook_patch.yaml` is enabled under `patches` section in `config/default/kustomization.yaml` file
+
+### Deploying 
+
+Now we have all our code changes and manifests in place, so let's deploy it to
+the cluster and test it out.
+
+Ensure that you have installed [cert-manager] `0.9.0+` version in your cluster. We have
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/tutorial.md]: We can point to out cert-manager doc, and we need to update it to `0.9.0+`

File: docs/book/src/multiversion-tutorial/testdata/project/api/v1/disk_types.go (hunk 1 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/testdata/project/api/v1/disk_types.go]: typo :) I meant `note` not `not`. Fixed it.

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+### Conversion Webhook
+API clients such as kubectl, controllers can request different versions of your
+API. So when a client requests for a version other than the storage version of
+your API, Kubernetes API server calls out to an HTTP endpoint to perform the
+conversion between the requested version and storage version. This HTTP endpoint
+is called Conversion Webhook and its discovery/connection parameters are
+configured in the CRD definition. With kubebuilder, you just need to implement
+conversion functions between different versions and it takes care of the rest of
+the work associated with running a webhook server, generating and pumbing the
```
[the remark this answers is not in the document]
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Done.

File: docs/book/src/multiversion-tutorial/tutorial.md (hunk 6 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Done.

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+configured in the CRD definition. With kubebuilder, you just need to implement
+conversion functions between different versions and it takes care of the rest of
+the work associated with running a webhook server, generating and pumbing the 
+conversion webhook handler.
+
+## Prerequisites
+CRD conversion webhook support was introduced as alpha feature in Kubernetes 1.13
+release. So ensure that you have a Kubernetes cluster that supports CRD conversion feature enabled.
+Refer to [instructions](../TODO.txt) to enable CRD conversion feature in your
+cluster.
```
[the remark this answers is not in the document]
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Good pt. Added.

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+conversion between the requested version and storage version. This HTTP endpoint
+is called Conversion Webhook and its discovery/connection parameters are
+configured in the CRD definition. With kubebuilder, you just need to implement
+conversion functions between different versions and it takes care of the rest of
+the work associated with running a webhook server, generating and pumbing the 
+conversion webhook handler.
+
+## Prerequisites
+CRD conversion webhook support was introduced as alpha feature in Kubernetes 1.13
+release. So ensure that you have a Kubernetes cluster that supports CRD conversion feature enabled.
```
[the remark this answers is not in the document]
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Good pt. Added.

File: docs/book/src/multiversion-tutorial/tutorial.md (hunk 8 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Added link to the cert-manager doc.

File: docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go (hunk 2 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/testdata/project/api/v2/disk_types.go]: I tried doing it, but it isn't working as expected. Let's chat offline to debug it.

File: docs/book/src/SUMMARY.md [hunk 12]
```diff
@@ -28,6 +28,10 @@
     - [Deploying webhooks](./cronjob-tutorial/running-webhook.md)
 
   - [Epilogue](./cronjob-tutorial/epilogue.md)
+- [Tutorial: Multi version API](./multiversion-tutorial/tutorial.md)
```
Comment (mengqiy) [inline docs/book/src/SUMMARY.md]: Should be `Multi-version`?

File: docs/book/src/multiversion-tutorial/tutorial.md [hunk 13]
```diff
@@ -1,303 +1,8 @@
 # Implement multi-version API
 
-This tutorial will teach you how to build an API with multiple versions.
+This tutorial will teach you how to build an API with multiple versions. We have
+decided the tutorial in three sections:
```
Comment (mengqiy) [inline docs/book/src/multiversion-tutorial/tutorial.md]: s/decided/divided ?

File: docs/book/src/multiversion-tutorial/testdata/project/config/default/kustomization.yaml (hunk 7 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/testdata/project/config/default/kustomization.yaml]: Done.

File: docs/book/src/multiversion-tutorial/tutorial.md (hunk 13 above)
Comment (droot) [reply docs/book/src/multiversion-tutorial/tutorial.md]: Done.

**Pull request merged**